README.md
# node-hledger

Node.js API for [hledger].

```
npm install --save-exact hledger
```

[![Status](https://travis-ci.org/rstacruz/node-hledger.svg?branch=master)](https://travis-ci.org/rstacruz/node-hledger "See test builds")

## API

<!--api-->

### hledger

> `hledger(args, options)`

Invokes hledger and returns a promise. It resolves into the CSV output as a
2D array.

```js
var hledger = require('hledger')

hledger(['bal', 'Assets'])
  .then((data) => ...)

[ [ 'account', 'balance' ],
  [ 'Assets', '$200' ],
  [ 'Assets:Savings', '$150' ],
  [ 'Assets:Checking', '$50' ] ]
```

You can invoke it with a string:

```js
hledger('bal Assets')
```

You can then use functions to make better sense of them:

```js
hledger(['bal', 'Assets'])
  .then(hledger.tableize)
  .then((data) => ...)

[ { account: 'Assets', balance: '$200' },
  { account: 'Assets:Savings', balance: '$150' },
  ... ]
```

You may pass the option `{ mode: 'list' }` to support commands that don't
have CSV output (eg, `accounts`).

```js
hledger('accounts', { mode: 'list' })
.then(data => ...)

[ 'Assets:Savings',
  'Assets:Checking'
]

```

### hledger.tableize

> `tableize(list)`

hledger.tableize:
Turns a CSV-based array into an table list.

```js
input = [
  ['account', 'amount'],
  ['Savings', '$100'],
  ['Checking', '$150']
]

tableize(input)
// [ { account: 'Savings', amount: '$100' },
//   { account: 'Checking', amount: '$200' } ]
```

Used for piping into `hledger()`'s promise output:

```js
hledger('...')
  .then(hledger.tableize)
  .then((data) => ...)
```
<!--api:end-->

[hledger]: http://hledger.org/

## Thanks

**node-hledger** © 2015+, Rico Sta. Cruz. Released under the [MIT] License.<br>
Authored and maintained by Rico Sta. Cruz with help from contributors ([list][contributors]).

> [ricostacruz.com](http://ricostacruz.com) &nbsp;&middot;&nbsp;
> GitHub [@rstacruz](https://github.com/rstacruz) &nbsp;&middot;&nbsp;
> Twitter [@rstacruz](https://twitter.com/rstacruz)

[MIT]: http://mit-license.org/
[contributors]: http://github.com/rstacruz/node-hledger/contributors


HISTORY.md
## [v1.1.1]
> Aug  8, 2016

- Support `{ mode: 'list' }` option.

[v1.1.1]: https://github.com/rstacruz/node-hledger/compare/v1.0.0...v1.1.1

## v1.0.0
> Dec 6, 2015

- Initial release.


test/index_test.js
var hl = require('../index')

describe('my project', function () {
  it('works', function () {
    return hl(['bal'])
      .then((data) => {
        console.log('return', JSON.stringify(data, null, 2))
      })
  })
})


test/lib/index_test.js
var hl = require('../../index')

describe('hledger', function () {
  describe('basic test', function () {
    before(function () {
      return hl(['-f', 'test/fixtures/test.ledger', 'bal'])
        .then(hl.tableize)
        .then((data) => {
          this.data = data
        })
    })

    it('parses 1st account', function () {
      expect(this.data[0])
        .toEqual({ account: 'Assets', balance: '$250' })
    })

    it('parses 2nd account', function () {
      expect(this.data[1])
        .toEqual({ account: 'Assets:Checking', balance: '$150' })
    })
  })

  describe('errors when hledger is not found', function () {
    before(function () {
      process.env.HLEDGER_BIN = 'nonexistentexecutable'
    })

    after(function () {
      delete process.env.HLEDGER_BIN
    })

    it('are handled', function () {
      return invert(hl(['...']))
      .then((err) => {
        expect(err.message).toEqual('spawn nonexistentexecutable ENOENT')
      })
    })
  })

  describe('handles accounts', function () {
    it('works', function () {
      return hl(['-f', 'test/fixtures/test.ledger', 'accounts'], { mode: 'list' })
      .then((data) => {
        expect(data).toEqual([
          'Assets:Checking',
          'Assets:Savings',
          'Equity:Opening balances'
        ])
      })
    })
  })

  describe('errors with unknown flags', function () {
    it('are handled', function () {
      return invert(hl(['-X']))
      .then((err) => {
        expect(err.message).toInclude('hledger: Unknown flag: -X')
      })
    })

    it('includes the code', function () {
      return invert(hl(['-X']))
      .then((err) => {
        expect(err.code).toEqual(1)
      })
    })
  })
})

/*
 * Inverts a promise
 */

function invert (promise) {
  return new Promise((resolve, reject) => {
    promise
      .then((data) => {
        reject(new Error('Promise succeeded when expected to fail'))
      })
      .catch((err) => {
        resolve(err)
      })
  })
}


test/lib/to_args_test.js
var toArgs = require('../../lib/to_args')

describe('to args', function () {
  it('works with arrays', function () {
    expect(toArgs(['a', 'b'])).toEqual(['a', 'b'])
  })

  it('works with strings', function () {
    expect(toArgs('a b')).toEqual(['a', 'b'])
  })

  it('works with strings with quotes', function () {
    expect(toArgs('a "b c" d')).toEqual(['a', 'b c', 'd'])
  })

  it('works with strings with control chars', function () {
    expect(toArgs('a "b c" > d')).toEqual(['a', 'b c', '>', 'd'])
  })
})


lib/index.js
'use strict'

var Promise = require('pinkie-promise')
var spawn = require('child_process').spawn
var csvParse = require('csv-parse')
var tableize = require('./tableize')
var toArgs = require('./to_args')

/**
 * Invokes hledger and returns a promise. It resolves into the CSV output as a
 * 2D array.
 *
 *     var hledger = require('hledger')
 *
 *     hledger(['bal', 'Assets'])
 *       .then((data) => ...)
 *
 *     [ [ 'account', 'balance' ],
 *       [ 'Assets', '$200' ],
 *       [ 'Assets:Savings', '$150' ],
 *       [ 'Assets:Checking', '$50' ] ]
 *
 * You can invoke it with a string:
 *
 *     hledger('bal Assets')
 *
 * You can then use functions to make better sense of them:
 *
 *     hledger(['bal', 'Assets'])
 *       .then(hledger.tableize)
 *       .then((data) => ...)
 *
 *     [ { account: 'Assets', balance: '$200' },
 *       { account: 'Assets:Savings', balance: '$150' },
 *       ... ]
 *
 * You may pass the option `{ mode: 'list' }` to support commands that don't
 * have CSV output (eg, `accounts`).
 *
 *     hledger('accounts', { mode: 'list' })
 *     .then(data => ...)
 *
 *     [ 'Assets:Savings',
 *       'Assets:Checking'
 *     ]
 */

function hledger (args, options) {
  var HLEDGER_BIN = process.env['HLEDGER_BIN'] || 'hledger'
  var mode = (options && options.mode) || 'csv'
  args = toArgs(args)

  if (mode === 'csv') {
    args = args.concat(['-O', 'csv'])
  }

  return new Promise(function (resolve, reject) {
    var child = spawn(HLEDGER_BIN, args)
    var stderr = ''
    var stdout = ''

    child.stdout.on('data', function (data) { stdout += data.toString() })
    child.stderr.on('data', function (data) { stderr += data.toString() })
    child.on('error', (err) => reject(err))

    child.on('close', function (code) {
      if (code !== 0) {
        var err = new Error(stderr.trim())
        err.code = code
        reject(err)
      } else if (mode === 'csv') {
        csvParse(stdout, function (err, data) {
          if (err) throw reject(err)
          resolve(data)
        })
      } else {
        resolve(stdout.trim().split('\n'))
      }
    })
  })
}

//
// Exports
//

hledger.tableize = require('./tableize')

module.exports = hledger
